python/ray/experimental/sgd/pytorch/pytorch_runner.py

```diff
@@ -0,0 +1,163 @@
+from __future__ import print_function
+
+import logging
+import os
+import torch
+import torch.distributed as dist
+import torch.utils.data
+
+import ray
+
+from ray.experimental.sgd.pytorch import utils
+
+logger = logging.getLogger(__name__)
+
+
+class PyTorchRunner(object):
+    """Manages a distributed PyTorch model replica.
+
+    Args:
+        model_creator (dict -> torch.nn.Module): creates the model using the
+            config.
+        data_creator (dict -> Dataset, Dataset): creates the training and
+            validation data sets using the config.
+        optimizer_creator (model, dict -> loss, optimizer): creates the loss
+            and optimizer using the config.
+        config (dict): configuration passed to 'model_creator', 'data_creator',
+            and 'optimizer_creator'.
+        batch_size (int): batch size used for SGD.
+        backend (string): backend used for distributed SGD. "gloo" or "nccl".
+    """
+
+    def __init__(self,
+                 model_creator,
+                 data_creator,
+                 optimizer_creator,
+                 config=None,
+                 batch_size=16,
+                 backend="gloo"):
+
+        self.model_creator = model_creator
+        self.data_creator = data_creator
+        self.optimizer_creator = optimizer_creator
+        self.config = {} if config is None else config
+        self.batch_size = batch_size
+        self.backend = backend
+        self.verbose = True
+
+        self.local_rank = None
+        if os.environ.get("CUDA_VISIBLE_DEVICES", None):
+            # TODO: might break if multiple GPUs requested
+            self.local_rank = int(os.environ["CUDA_VISIBLE_DEVICES"])
+        self.epoch = 0
+        self._timers = {
+            "setup_proc": utils.TimerStat(window_size=1),
+            "setup_model": utils.TimerStat(window_size=1),
+            "get_state": utils.TimerStat(window_size=1),
+            "set_state": utils.TimerStat(window_size=1),
+            "validation": utils.TimerStat(window_size=1),
+            "training": utils.TimerStat(window_size=1)
+        }
+
+    def setup(self, url, world_rank, world_size):
+        self.setup_distributed_pytorch(url, world_rank, world_size)
+        self.setup_training()
+
+    def setup_distributed_pytorch(self, url, world_rank, world_size):
+        os.environ["CUDA_LAUNCH_BLOCKING"] = "1"
+        with self._timers["setup_proc"]:
+            self.world_rank = world_rank
+            logger.debug(
+                "Connecting to {} world_rank: {} world_size: {}".format(
+                    url, world_rank, world_size))
+            logger.debug("using {}".format(self.backend))
+            dist.init_process_group(backend=self.backend,
+                                    init_method=url,
+                                    rank=world_rank,
+                                    world_size=world_size)
+
+            # This is a hack because set_devices fails otherwise
```

Comment:
It'd be nice to raise an issue on the PyTorch side for this?

Reply:
I took this from your implementation. I'm happy to make the issue, but you might be better positioned?

Reply:
I think you should do it. We should leave this here, but I suspect this will raise issues, and we should aim to work with them to remove this hack as soon as possible.

Reply:
Turns out this hack isn't needed since PyTorch autodetects the device from `CUDA_AVAILABLE_DEVICES`, and Ray sets this correctly.

Reply:
Also, I'm not sure if this is a PyTorch bug if `set_devices` operates on `CUDA_VISIBLE_DEVICES`

Reply:
Oh this is hard to test corectly. You'll need to run 2 SGD instances for this error to show - PyTorch requires CUDA_AVAILABLE_DEVICES to be contiguous, and Ray does not guarantee that.

Reply:
I'm not quite sure I understand the error then...the current implementation seems to work on a multi-GPU instance without the hack. My impression is that this is a workaround  for not being able to request GPUs for the runner actor. Is it even necessary now that that runners can request GPUs?

Reply:
Can you try 3 SGD instantiations on one instance - 1 gpu, 2 gpu, 1 gpu (in that order) on a 4 gpu machine? There’s a particular error that is causes when CUA